### PR TITLE
Certificate handler methods work with generic certificate type.

### PR DESCRIPTION
**Pull request by deuszx**
## Motivation

Another PR on the road to complete removal of `CertificateValue` and (old) `Certificate` types.

## Proposal

In this PR we mostly update the `handle_certificate` methods to work over a generic `GenericCertificate<T>` or create handles for new certificate types.

## Test Plan

CI.

## Release Plan

- Nothing to do / These changes follow the usual release cycle.

## Links

Closes https://github.com/linera-io/linera-protocol/issues/2967

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-chain/src/certificate/hashed.rs [hunk 1]
```diff
@@ -102,6 +108,60 @@ impl<T> PartialEq for Hashed<T> {
...
+    fn get(&self) -> R;
+}
+
+pub enum IsValidated {}
+
+impl Has<IsValidated, bool> for ValidatedBlock {
+    fn get(&self) -> bool {
+        true
+    }
+}
```
Comment (ma2bd) [inline linera-chain/src/certificate/hashed.rs:121]: This `Has` trait is starting to look very suspicious.

Comment (deuszx) [reply linera-chain/src/certificate/hashed.rs:121]: Why? I think it's great (although I'm biased of course) way to ensure that a type `T` has some specific getter.

Comment (deuszx) [reply linera-chain/src/certificate/hashed.rs:121]: I don't understand what difference does it make 🤔 it many places the values we get out of the getters are used only before they're signed. And once they're part of the certificate, the getters are not used anymore.

Comment (deuszx) [reply linera-chain/src/certificate/hashed.rs:121]: And actually I think the previous way it was done was incorrect - the `CertificateValue` was a "catch-it-all" struct that included chainid, round, epoch but also required blob ids, whether value was of a certain type (`is_validated`). It wasn't "one thing", it was many things.

Comment (afck) [reply linera-chain/src/certificate/hashed.rs:121]: I'd say it makes at least a difference in terms of how understandable the code is: "This type can end up as the value of a certificate" tells me more than "this type has a chain ID".

Comment (deuszx) [reply linera-chain/src/certificate/hashed.rs:121]: I'd say that the point of working with generic functions is to require as little as possible from the caller. If I were to replace this approach with something like a `trait CertificateValueT` (working name) and require that instead then some places will require unnecessary constraints from the arguments and the implementations for these types will have dummy values that don't make sense (just like it was before).

I'll do the change and we'll judge again.

Comment (afck) [reply linera-chain/src/certificate/hashed.rs:121]: If we end up in that situation it just means we applied the `CertificateValueT` trait to something that actually isn't a certificate value, just to get a chain ID getter.

Are there really any examples in the code where we handle "anything with a chain ID" but it's not necessarily a certificate value?

If we do, I'd find that suspicious, but if we really need it, we could always `impl<T> Has<ChainId> for T where T: CertificateValueT`.

Comment (deuszx) [reply linera-chain/src/certificate/hashed.rs]: Done in https://github.com/linera-io/linera-protocol/pull/2921/commits/db80ea5d48dfef41494414d91dcff47e1b2cd1ab . I think you might have been right and it does _read_ better.

Comment (deuszx) [reply linera-chain/src/certificate/hashed.rs]: Just FYI @ma2bd  - I'm resolving this comment as `Has` trait has been removed entirely from the codebase.

File: linera-core/src/local_node.rs [hunk 2]
```diff
@@ -110,29 +110,34 @@ where
...
-            notifier,
-        ))
-        .await?;
-        Ok(response)
+    ) -> Result<ChainInfoResponse, LocalNodeError>
+    where
+        T: CertificateProcessor,
+    {
+        Ok(
+            Box::pin(self.node.state.fully_handle_certificate_with_notifications(
```
Comment (bart-linera) [inline linera-core/src/local_node.rs:134]: Is the `pin` still necessary here? Looks like it shouldn't be to me...

File: linera-chain/src/certificate/generic.rs [hunk 3]
```diff
@@ -99,6 +99,18 @@ impl<T> GenericCertificate<T> {
         crate::data_types::check_signatures(self.hash(), self.round, &self.signatures, committee)?;
         Ok(())
     }
+
+    pub fn lite_cerificate(&self) -> crate::certificate::LiteCertificate<'_>
```
Comment (bart-linera) [inline linera-chain/src/certificate/generic.rs:103]: Typo in `lite_certificate` (missing a `t`) :sweat_smile:

File: linera-core/src/local_node.rs (hunk 2 above)
Comment (bart-linera) [reply linera-core/src/local_node.rs]: Whoops, looks like it was needed after all :grimacing: (CI threw some errors because of its removal.)

Comment (deuszx) [reply linera-core/src/local_node.rs:134]: Yeah... brought back.

File: linera-core/src/worker.rs [hunk 4]
```diff
@@ -761,85 +824,131 @@ where
...
+                    notify_when_messages_are_delivered,
+                )
+                .await
+            }
+            Either::Right(validated) => self.handle_validated_certificate(validated, vec![]).await,
+        }
     }
 
-    /// Processes a certificate.
+    /// Processes a block certificate.
```
Comment (afck) [inline linera-core/src/worker.rs:840]: But it does also process a timeout certificate.
("Block certificate" reads like "validated or confirmed", i.e. one that contains a block?)

File: linera-core/src/worker.rs [hunk 5]
```diff
@@ -761,85 +824,131 @@ where
...
     #[instrument(skip_all, fields(
-        nick = self.nickname,
-        chain_id = format!("{:.8}", certificate.inner().chain_id()),
-        height = %certificate.inner().height(),
-    ))]
+            nick = self.nickname,
+            chain_id = format!("{:.8}", certificate.inner().chain_id()),
+            height = %certificate.inner().height(),
+        ))]
```
Comment (afck) [inline linera-core/src/worker.rs:845]: That reformatting looks wrong.

File: linera-core/src/worker.rs [hunk 6]
```diff
@@ -761,85 +824,131 @@ where
...
+                .await
             }
-            CertificateValue::Timeout { .. } => {
-                // Handle the leader timeout.
-                // Note: This conversion panics if `certificate` is not a timeout certificate
-                // but we just checked that it is.
-                let timeout_certificate = certificate.into();
-                self.process_timeout(timeout_certificate).await?
+            CertificateValue::ValidatedBlock(_) => {
+                let certificate = ValidatedBlockCertificate::from(certificate);
```
Comment (afck) [inline linera-core/src/worker.rs:863]: For `ConfirmedBlock` we use `try_from(…).unwrap()`. Would be good to be consistent. (Same for `Timeout` below.)

Is the plan to replace `Certificate` with an enum that contains the different types of `GenericCertificate`s? That would make the `unwrap`s here unnecessary.

File: linera-core/src/worker.rs [hunk 7]
```diff
@@ -761,85 +824,131 @@ where
...
+        blobs: Vec<Blob>,
+    ) -> Result<(ChainInfoResponse, NetworkActions), WorkerError> {
+        trace!("{} <-- {:?}", self.nickname, certificate);
+
+        #[cfg(with_metrics)]
+        let round = certificate.round;
+        #[cfg(with_metrics)]
+        let cert_str = certificate.inner().to_log_str();
+
+        let (info, actions, _duplicated) =
```
Comment (afck) [inline linera-core/src/worker.rs:927]: (Not in this PR, but: Now I'm wondering why we don't check for duplicates in the other cases.)

File: linera-core/src/worker.rs [hunk 8]
```diff
@@ -761,85 +824,131 @@ where
...
+        chain_id = format!("{:.8}", certificate.inner().chain_id),
+        height = %certificate.inner().height,
+    ))]
+    pub async fn handle_timeout_certificate(
+        &self,
+        certificate: TimeoutCertificate,
+    ) -> Result<(ChainInfoResponse, NetworkActions), WorkerError> {
+        trace!("{} <-- {:?}", self.nickname, certificate);
+        let (info, actions) = self.process_timeout(certificate).await?;
         Ok((info, actions))
```
Comment (afck) [inline linera-core/src/worker.rs:952]:
```suggestion
        self.process_timeout(certificate).await
```

(Should that function be inlined here?)

**Review by afck — COMMENTED** [review]
Looks great! But the PR needs a description.

File: linera-chain/src/certificate/value.rs [hunk 9]
```diff
@@ -139,29 +139,37 @@ pub type HashedCertificateValue = Hashed<CertificateValue>;
...
 impl HashedCertificateValue {
     /// Creates a [`ConfirmedBlock`](CertificateValue::ConfirmedBlock) value.
-    pub fn new_confirmed(executed_block: ExecutedBlock) -> HashedCertificateValue {
-        CertificateValue::ConfirmedBlock(ConfirmedBlock::new(executed_block)).into()
+    pub fn new_confirmed(executed_block: ExecutedBlock) -> Hashed<ConfirmedBlock> {
+        HashedCertificateValue::from(CertificateValue::ConfirmedBlock(ConfirmedBlock::new(
+            executed_block,
+        )))
+        .try_into()
+        .unwrap()
```
[the remark this answers is not in the document]
Comment (deuszx) [reply linera-chain/src/certificate/value.rs:147]: Not yet, b/c we still verify the hash over `HashedCertificateValue` and not `Hashed<ConfirmedBlock>`.

File: linera-core/src/client/mod.rs [hunk 10]
```diff
@@ -2276,21 +2265,6 @@ where
...
-        let next_block_height = self.next_block_height();
-        self.communicate_chain_updates(
-            &committee,
-            self.chain_id,
-            next_block_height,
-            CrossChainMessageDelivery::NonBlocking,
-        )
-        .await?;
-        Ok(())
-    }
```
[the remark this answers is not in the document]
Comment (deuszx) [reply linera-core/src/client/mod.rs:2292]: If you look closely I had to change the flow for block proposal a bit and extracted the "update validators" to a new function due to "type constraints" (i.e. we have `GenericCertificate<T>` while the `finalize_vote` can be called only on `GenericCertificate<ValidatedBlock>`). So there were now two methods `update_validators` with a very similar body.

File: linera-chain/src/certificate/value.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (afck) [reply linera-chain/src/certificate/value.rs:147]: Yes, but we should probably change that in the future?

[the remark this answers is not in the document]
Comment (deuszx) [reply linera-chain/src/certificate/value.rs:147]: Yes, in the (near) future the `HashedCertificateValue` will be removed entirely.

File: linera-core/src/worker.rs [hunk 11]
```diff
@@ -396,51 +398,112 @@ where
...
-                .map(Into::into)?),
+                .map(Either::Right)?),
             Err(other) => Err(other),
         }
     }
 }
 
+#[allow(async_fn_in_trait)]
+#[cfg_attr(not(web), trait_variant::make(Send))]
+pub trait CertificateProcessor: CertificateValueT + Sized + 'static {
```
[the remark this answers is not in the document]
Comment (deuszx) [reply linera-core/src/worker.rs:420]: It processes a certificate so a noun becomes `CertificateProcessor` 🤔

File: linera-core/src/worker.rs (hunk 6 above)
Comment (deuszx) [reply linera-core/src/worker.rs:863]: Yes, the plan is to put a `typ` field in the `api::Certificate` definition (protobuf).

Comment (deuszx) [reply linera-core/src/worker.rs:863]: but I'm trying different approaches.

File: linera-core/src/worker.rs (hunk 4 above)
Comment (deuszx) [reply linera-core/src/worker.rs:840]: I'm gonna rm that word from here.

File: linera-core/src/unit_tests/worker_tests.rs [hunk 12]
```diff
@@ -716,14 +720,15 @@ where
             .unwrap()
             .value()
             .inner()
+            .inner()
```
[the remark this answers is not in the document]
Comment (deuszx) [reply linera-core/src/unit_tests/worker_tests.rs]: Done in [Rename inner to executed_block](https://github.com/linera-io/linera-protocol/pull/2921/commits/06ad449bdaef1938304352353aaa85a137831a7c).

File: linera-core/src/worker.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (afck) [reply linera-core/src/worker.rs:420]: I'd say it _is_ a certificate, and it knows how to pass itself to a worker.

From a `CertificateProcessor` I would expect a function like `fn process_certificate(&self, certificate: …)`.

[the remark this answers is not in the document]
Comment (deuszx) [reply linera-core/src/worker.rs:420]: Ok, I'll rename to `ProcessableCertificate`.

**Review by afck — APPROVED** [review]

**Pull request merged**